components/raftstore/src/store/peer.rs

```diff
@@ -2185,22 +2177,37 @@ where
             "region_id" => self.region_id,
             "peer_id" => self.peer.get_id(),
             "request" => ?change_peer,
-            "total" => total,
-            "after" => progress.voter_ids().len(),
+            "before" => ?before_progress.conf().to_conf_state(),
+            "after" => ?after_progress.conf().to_conf_state(),
             "truncated_index" => self.get_store().truncated_index(),
             "promoted_commit_index" => promoted_commit_index,
         );
         // Waking it up to replicate logs to candidate.
         self.should_wake_up = true;
         Err(box_err!(
-            "unsafe to perform conf change {:?}, total {}, truncated index {}, promoted commit index {}",
+            "unsafe to perform conf change {:?}, truncated index {}, promoted commit index {}",
             change_peer,
-            total,
             self.get_store().truncated_index(),
             promoted_commit_index
         ))
     }
 
+    /// Check if current joint state can handle this confchange
+    fn check_joint_state(&mut self, cc: &impl ConfChangeI) -> Result<ProgressTracker> {
```

Comment:
This logic is a little duplicated in raft-rs

Reply:
Yes, it is, that is because the check is inside `raft-rs`'s `apply_conf_change` which is called after the confchange log was applied, but here we want to do the check early.